fix #142

aviatesk opened the pull request:
EDIT: please see https://github.com/JuliaDebug/Cthulhu.jl/pull/144#issuecomment-812307339 for the actual changes this PR will make

By making sure `slottypes::Vector{Any}`

@vtjnash could you give me a MRE to reproduce #142 ? I'm still not sure why/how `CodeInfo.slottypes` can be `nothing`.

vtjnash commented:
sorry, this isn't a great example, since it is quite slow (and seems like Cthulhu no longer has the ability to cache between calls), but this is what I was trying to debug:
```
julia> using MixedModels: dataset                                                                
                                                                                                                                                                                                   
julia> using MixedModels                                                                                                                                                                           
                                                                                                 
julia> form = @formula((incid/hsz) ~ 1 + period + (1|herd))                                      
FormulaTerm                                                                                                                                                                                        
Response:                                                                                        
  (incid,hsz)->incid / hsz                                                                                                                                                                         Predictors:                                                                                                                                                                                          1                                   
  period(unknown)                     
  (herd)->1 | herd                    
                                                
julia> @descend GeneralizedLinearMixedModel(form, cbpp,                                          
                                                  Binomial();                                    
                                                  wts=float(cbpp.hsz))
> select 1
ERROR: KeyError: key MethodInstance for (::Core.var"#Type##kw")(::NamedTuple{(:wts,), Tuple{Vector{Float64}}}, ::Type{GeneralizedLinearMixedModel}, ::StatsModels.FormulaTerm{StatsModels.FunctionTerm{typeof(/), var"#3#5", (:incid, :hsz)}, Tuple{StatsModels.ConstantTerm{Int64}, StatsModels.Term, StatsModels.FunctionTerm{typeof(|), var"#4#6", (:herd,)}}}, ::Arrow.Table, ::Binomial{Float64}, ::LogitLink) not found

> select o
> select 2
ERROR: MethodError: no method matching argextype(::Core.SSAValue, ::Core.CodeInfo, ::Vector{Any}, ::Nothing)
```

aviatesk commented:
Thanks, and now the example will produce the same error as reported in https://github.com/JuliaLang/julia/issues/40048 ...
```julia
julia> using MixedModels: dataset; using MixedModels; form = @formula((incid/hsz) ~ 1 + period + (1|herd)); cbpp = dataset(:cbpp);

julia> @descend GeneralizedLinearMixedModel(form, cbpp,                                          
                                                  Binomial();                                    
                                                  wts=float(cbpp.hsz))
ERROR: StackOverflowError:
...
```

simeonschaub commented:
A more self-contained example:
```julia
julia> using Cthulhu

julia> @descend mapreduce(identity, vcat, ([5], [1.]))

│ ─ %-1  = invoke mapreduce(::typeof(identity),::typeof(vcat),::Tuple{Vector{Int64}, Vector{Float64}})::AbstractVector{T} where T
289 1 ─ %1 = invoke Base.:(var"#mapreduce#237")($(QuoteNode(Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}()))::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, _1::typeof(mapreduce), _2::Function, _3::Function, _4::Tuple{Vector{Int64}, Vector{Float64}})::AbstractVector{T} where T
    └──      return %1                                                                                          │

Select a call to descend into or ↩ to ascend. [q]uit. [b]ookmark.
Toggles: [o]ptimize, [w]arn, [v]erbose printing for warntype code, [d]ebuginfo, [s]yntax highlight for Source/LLVM/Native.
Show: [S]ource code, [A]ST, [L]LVM IR, [N]ative code
Actions: [E]dit source code, [R]evise and redisplay
Advanced: dump [P]arams cache.
 • %1  = invoke #mapreduce#237(…,…,…,…,…)::AbstractVector{T} where T
   ↩
ERROR: KeyError: key MethodInstance for Base.var"#mapreduce#237"(::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}}, ::typeof(mapreduce), ::typeof(identity), ::typeof(vcat), ::Tuple{Vector{Int64}, Vector{Float64}}) not found
Stacktrace:
 [1] getindex
   @ ./dict.jl:482 [inlined]
 [2] _descend(interp::Cthulhu.CthulhuInterpreter, mi::Core.MethodInstance; override::Nothing, iswarn::Bool, params::Core.Compiler.NativeInterpreter, optimize::Bool, interruptexc::Bool, verbose::Bool, kwargs::Base.Pairs{Symbol, Symbol, Tuple{Symbol}, NamedTuple{(:debuginfo,), Tuple{Symbol}}})
   @ Cthulhu ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:208
 [3] _descend(interp::Cthulhu.CthulhuInterpreter, mi::Core.MethodInstance; override::Nothing, iswarn::Bool, params::Core.Compiler.NativeInterpreter, optimize::Bool, interruptexc::Bool, verbose::Bool, kwargs::Base.Pairs{Symbol, Union{}, Tuple{}, NamedTuple{(), Tuple{}}})
   @ Cthulhu ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:295
 [4] _descend(F::Any, TT::Any; params::Core.Compiler.NativeInterpreter, kwargs::Base.Pairs{Symbol, Bool, Tuple{Symbol}, NamedTuple{(:iswarn,), Tuple{Bool}}})
   @ Cthulhu ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:373
 [5] _descend_with_error_handling(::Any, ::Vararg{Any}; kwargs::Base.Pairs{Symbol, V, Tuple{Vararg{Symbol, N}}, NamedTuple{names, T}} where {V, N, names, T<:Tuple{Vararg{Any, N}}})
   @ Cthulhu ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:133
 [6] #descend_code_typed#46
   @ ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:105 [inlined]
 [7] descend_code_typed(f::Function, tt::Any)
   @ Cthulhu ~/.julia/packages/Cthulhu/7xpUc/src/Cthulhu.jl:105
 [8] top-level scope
   @ REPL[2]:1
```
Doesn't seem like this PR fixes this particular example though.

simeonschaub commented on src/Cthulhu.jl:
```diff
@@ -174,15 +174,18 @@ function lookup(interp::CthulhuInterpreter, mi::MethodInstance, optimize::Bool)
         infos = interp.unopt[mi].stmt_infos
         slottypes = codeinf.slottypes
+        if isnothing(slottypes)
+            slottypes = Any[ Any for i = 1:nslots ]
```
`nslots` does not seem to be defined anywhere

aviatesk replied:
ops I just forgot to push the local fix. Thanks.

vchuravy commented:
Great work! Is it possible to write a test for this!

aviatesk commented:
Okay, so the problem was kinda deeply nested than I expected.

The first thing to notice is that there are cases when we can't find a
cache in `CthulhuInterpreter.opt`.
AFAIU this is mainly because of limited frames; the native interpreter
doesn't cache them but there are still statement infos about them and
so the problem happens if a Cthulhu user tries to inspect it.

When trying to inspect uncached frames, the next problem is how to
handle them at pre-optimization phase. The `slottypes` of them would be
`nothing` at that point, and so we need to do something like
[what the external `OptimizationState` constructor does](https://github.com/JuliaLang/julia/blob/878e1cd52ebf14e8666d8d54a54a95ccee4405d3/base/compiler/optimize.jl#L69-L71).
The another problem is the return types of those frames can be annotated
as `LimitedAccuracy` at that point, and it contains recursive frame chain
and thus the printer can fail into infinite loop. The same issue happens
when Cthulhu tries to show `CodeInfo` object, which may also contain
`LimitedAccuracy`.

Now this commit addresses the issues outlined above:
- check if a callsite is cached or not, and wrap it into `UncachedCallInfo` if not,
  and Cthulhu doesn't recurse into the `UncachedCallInfo` (or we can even
  forcibly recurse into, but I decided not to do that, since it's really
  different behaviour from what the native interpreter did).
  Note that `UncachedCallInfo` won't happen when working on a pre-optimization state,
  since Cthulhu will cache unoptimized frames by hooking into
  `finish(::InferenceState, ::CthulhuInterpreter)` in that case,
  which caches a frame no matter whether it is limited or not.
- when working on a pre-optimization state, if the return type of a callsite
  is annotated as `LimitedAccuracy`, we widen it and wrap it into
`LimitedCallInfo`.
  When printing `LimitedCallInfo`, Cthulhu only indicates the call is limited,
  but won't try to print the whole `LimitedAccuracy` object.
- when working on pre-optimization state, we make sure we don't print `LimitedAccuracy`
  included in `CodeInfo`. Currently I wrote an naive hack to just copy
  the original `CodeInfo` and widen `LimitedAccuracy`s in `ssavaluetypes` and `rettype`.

The changes above will fix #142 as well as [the case @simeonschaub reporeted here](https://github.com/JuliaDebug/Cthulhu.jl/pull/144#issuecomment-810959321).

---

@Keno (@vtjnash) do these changes sound fine ?

vchuravy commented:
Thanks @aviatesk! Can you rebase?

aviatesk commented:
Done :)

simeonschaub commented on test/runtests.jl:
```diff
@@ -118,9 +123,30 @@ let callsites = find_callsites_by_ftt(f_matches, Tuple{Any, Any}; optimize=false
-@testset "union-split constant-prop'ed callsites" begin
-    anonymous_module() = Core.eval(@__MODULE__, :(module $(gensym()) end))::Module
+# NOTE some inference won't work with inner functions
+anonymous_module() = Core.eval(@__MODULE__, :(module $(gensym()) end))::Module
```
Just FYI, I think you can just use `Module()` for this, but doesn't really matter here.

aviatesk replied:
That's neat. Thanks.

vchuravy commented on src/Cthulhu.jl:
```diff
@@ -277,7 +282,18 @@ function _descend(interp::CthulhuInterpreter, mi::MethodInstance; override::Unio
+            if info isa UncachedCallInfo
+                @warn "Cthulhu can't recurse into the uncached callsite"
+                continue
+                # # or, we can forcibly enter into the frame, which the native interpreter threw away
```
I suspect we would want to do that. Otherwise it will be hard for the user to investigate further from this point.

aviatesk replied:
Yeah, that's also a valid argument. Applied with some other minor fixes: 9624b3b

aviatesk commented:
Let me merge this within a day.

aviatesk commented:
Going to merge.
